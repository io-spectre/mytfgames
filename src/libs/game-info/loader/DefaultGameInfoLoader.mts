import { promises as fsa } from "node:fs";
import path from "node:path";

import { inject, injectable } from "inversify";

import {
  BundledArtifactInfo,
  BundledGameInfo,
  BundledVersionInfo,
  GameInfoParser,
} from "$game-info/parser";

import { BundledGameMetadata } from "./GameInfoLoader.mjs";

const META_FOLDER_NAME = ".meta";
const CHANGELOGS_FOLDER_NAME = "changelogs";
const GAME_MD_FILENAME = "game.md";

@injectable()
export class DefaultGameInfoLoader {
  constructor(
    @inject(GameInfoParser) private readonly gameInfoParser: GameInfoParser,
  ) {}

  async loadGameInfoFromMetaFolder(
    artifactFolder: string,
  ): Promise<BundledGameMetadata> {
    const metaFolderPath = path.join(artifactFolder, META_FOLDER_NAME);
    const gameMdPath = path.join(metaFolderPath, GAME_MD_FILENAME);
    const changelogsDirPath = path.join(metaFolderPath, CHANGELOGS_FOLDER_NAME);

    const [gameInfo, versionInfo] = await Promise.all([
      this.loadBundledGameInfo(gameMdPath),
      this.loadAllBundledVersionInfo(changelogsDirPath),
    ]);
    return { gameInfo, versionInfo };
  }

  async loadBundledGameInfo(gameMdPath: string): Promise<BundledGameInfo> {
    const gameMD = await fsa.readFile(gameMdPath, "utf-8");
    return this.gameInfoParser.parseGameInfo(gameMD);
  }

  async loadBundledArtifactInfo(
    artifactYamlPath: string,
  ): Promise<BundledArtifactInfo> {
    const artifactYaml = await fsa.readFile(artifactYamlPath, "utf-8");
    return this.gameInfoParser.parseArtifactYaml(artifactYaml);
  }

  async loadAllBundledVersionInfo(
    changelogsDirPath: string,
  ): Promise<BundledVersionInfo[]> {
    const changelogFiles = await fsa.readdir(changelogsDirPath, {
      withFileTypes: true,
      recursive: false,
    });
    const versionInfos = await Promise.all(
      changelogFiles.map(async (changelogDirent) => {
        if (!changelogDirent.isFile()) {
          return null;
        }
        const changelogFilename = changelogDirent.name;
        if (
          changelogFilename.length < 4 || // require at least one version character
          changelogFilename.slice(-3).toLowerCase() !== ".md"
        ) {
          return null;
        }

        const version = changelogFilename.slice(0, -3);
        try {
          return await this.loadBundledVersionInfo(
            version,
            path.join(changelogsDirPath, changelogFilename),
          );
        } catch {
          return null;
        }
      }),
    );
    return versionInfos.filter((v) => v != null);
  }
  async loadBundledVersionInfo(version: string, changelogMdPath: string) {
    const changelogMD = await fsa.readFile(changelogMdPath, "utf-8");
    return this.gameInfoParser.parseChangelog(version, changelogMD);
  }
}
