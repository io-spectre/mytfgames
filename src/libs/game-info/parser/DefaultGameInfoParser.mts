import parseFrontMatter from "gray-matter";
import { inject, injectable } from "inversify";
import { load as yamlLoad } from "js-yaml";
import * as uuid from "uuid";

import { Ajv } from "$node-base/utils";

import {
  ArtifactMetadataType,
  ArtifactMetadataTypeDefinition,
  ArtifactMetadataValidationFunction,
  ChangelogFrontMatterType,
  ChangelogFrontMatterTypeDefinition,
  ChangelogFrontMatterValidationFunction,
  GameInfoFrontMatterType,
  GameInfoFrontMatterTypeDefinition,
  GameInfoFrontMatterValidationFunction,
} from "./GameInfoJTDs.mjs";
import {
  BundledArtifactInfo,
  BundledAuthorInfo,
  BundledGameInfo,
  BundledVersionInfo,
  GameInfoParser,
} from "./GameInfoParser.mjs";

// note that the parser explicitly re-creates all objects and arrays while
// explicitly naming the properties to ensure that no extra properties are
// present in the parsed data

@injectable()
export class DefaultGameInfoParser implements GameInfoParser {
  readonly #validateGameInfoFrontMatter: GameInfoFrontMatterValidationFunction;
  readonly #validateArtifactMetadata: ArtifactMetadataValidationFunction;
  readonly #validateChangelogFrontMatter: ChangelogFrontMatterValidationFunction;

  constructor(@inject(Ajv) private readonly ajv: Ajv) {
    this.#validateGameInfoFrontMatter =
      this.ajv.compile<GameInfoFrontMatterType>(
        GameInfoFrontMatterTypeDefinition,
      );
    this.#validateArtifactMetadata = this.ajv.compile<ArtifactMetadataType>(
      ArtifactMetadataTypeDefinition,
    );
    this.#validateChangelogFrontMatter =
      this.ajv.compile<ChangelogFrontMatterType>(
        ChangelogFrontMatterTypeDefinition,
      );
  }

  parseGameInfo(gameMD: string): BundledGameInfo {
    const parsed = parseFrontMatter(gameMD, {
      language: "yaml",
    });

    if (!this.#validateGameInfoFrontMatter(parsed.data)) {
      throw new Error("Invalid game metadata", {
        cause: this.#validateGameInfoFrontMatter.errors,
      });
    }
    let gameUuid: Buffer;
    try {
      gameUuid = uuid.parse(parsed.data.ids.uuid) as Buffer;
    } catch (error) {
      throw new Error("Invalid game UUID", { cause: error });
    }

    const authors = parsed.data.authors.map((author, authorIndex) => {
      let authorUuid: Buffer;
      try {
        authorUuid = uuid.parse(author.uuid) as Buffer;
      } catch (error) {
        throw new Error(`Invalid author UUID at index ${authorIndex}`, {
          cause: error,
        });
      }
      return {
        metadataVersion: author["metadata-version"],
        uuid: authorUuid,
        name: author.name,
        tfgamesSiteProfileId: author["tfgames.site"] ?? null,
      } satisfies BundledAuthorInfo;
    });

    return {
      metadataVersion: parsed.data["metadata-version"],
      name: parsed.data.name,
      synopsis: parsed.data.synopsis,
      description: parsed.content,
      ids: {
        uuid: gameUuid,
        tfgamesSiteGameId: parsed.data.ids["tfgames.site"] ?? null,
      },
      authors,
      tags: parsed.data.tags,
    };
  }

  parseArtifactYaml(artifactYaml: string): BundledArtifactInfo {
    const parsed = yamlLoad(artifactYaml);

    if (!this.#validateArtifactMetadata(parsed)) {
      throw new Error("Invalid artifact metadata", {
        cause: this.#validateArtifactMetadata.errors,
      });
    }

    return {
      version: parsed.version,
      platform: parsed.platform,
    };
  }

  parseChangelog(version: string, changelogMD: string): BundledVersionInfo {
    const parsed = parseFrontMatter(changelogMD, {
      language: "yaml",
    });

    if (!this.#validateChangelogFrontMatter(parsed.data)) {
      throw new Error(`Invalid changelog metadata for "${version}"`, {
        cause: this.#validateChangelogFrontMatter.errors,
      });
    }

    return {
      version,
      note: parsed.content,
    };
  }
}
