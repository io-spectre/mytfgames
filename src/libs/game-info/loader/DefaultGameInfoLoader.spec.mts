import * as path from "node:path";

import { Ajv } from "ajv/dist/jtd";
import { beforeEach, describe, expect, it } from "vitest";

import { DefaultGameInfoParser } from "../parser/DefaultGameInfoParser.mjs";
import { DefaultGameInfoLoader } from "./DefaultGameInfoLoader.mjs";

describe("DefaultGameInfoLoader", () => {
  const ajv = new Ajv({ timestamp: "string" });
  let gameInfoParser: DefaultGameInfoParser;

  beforeEach(() => {
    gameInfoParser = new DefaultGameInfoParser(ajv);
  });

  it("should be instantiatable", () => {
    const subject = new DefaultGameInfoLoader(gameInfoParser);
    expect(subject).toBeInstanceOf(DefaultGameInfoLoader);
  });

  describe("instance", () => {
    let subject: DefaultGameInfoLoader;

    beforeEach(() => {
      subject = new DefaultGameInfoLoader(gameInfoParser);
    });

    it("should load game info from the meta folder", async () => {
      const gameMdPath = path.join(import.meta.dirname, "test/meta-0/game.md");

      const gameInfo = subject.loadBundledGameInfo(gameMdPath);
      await expect(gameInfo).resolves.matchSnapshot();
    });
    it("should throw an error when loading game info from a non-existent path", async () => {
      const gameMdPath = path.join(
        import.meta.dirname,
        "test/meta-0/does-not-exist.md",
      );

      const gameInfo = subject.loadBundledGameInfo(gameMdPath);
      await expect(gameInfo).rejects.toMatchObject({
        code: "ENOENT",
      });
    });

    it("should load artifact info from the meta folder", async () => {
      const artifactYamlPath = path.join(
        import.meta.dirname,
        "test/meta-0/artifact.yaml",
      );

      const artifactInfo = subject.loadBundledArtifactInfo(artifactYamlPath);
      await expect(artifactInfo).resolves.matchSnapshot();
    });
    it("should throw an error when loading artifact info from a non-existent path", async () => {
      const artifactYamlPath = path.join(
        import.meta.dirname,
        "test/meta-0/does-not-exist.yaml",
      );

      const artifactInfo = subject.loadBundledArtifactInfo(artifactYamlPath);
      await expect(artifactInfo).rejects.toMatchObject({
        code: "ENOENT",
      });
    });

    it("should load all bundled version info from the changelogs folder", async () => {
      const changelogsDirPath = path.join(
        import.meta.dirname,
        "test/meta-0/changelogs",
      );

      const versionInfos = subject.loadAllBundledVersionInfo(changelogsDirPath);
      await expect(versionInfos).resolves.matchSnapshot();
    });
    it("should throw an error when loading version info from a non-existent path", async () => {
      const changelogsDirPath = path.join(
        import.meta.dirname,
        "test/meta-0/does-not-exist",
      );

      const versionInfos = subject.loadAllBundledVersionInfo(changelogsDirPath);
      await expect(versionInfos).rejects.toMatchObject({
        code: "ENOENT",
      });
    });
  });
});
