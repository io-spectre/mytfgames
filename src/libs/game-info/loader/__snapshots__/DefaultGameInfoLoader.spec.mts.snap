// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DefaultGameInfoLoader > instance > should load all bundled version info from the changelogs folder 1`] = `
[
  {
    "note": "
# 1.0.3.0 "This And That"

## Play It Again Samantha!

- Option to replay certain 1-shot scenes.
- WIP scenes can now be replayed (look for a journal in your bedroom to trigger).

## Goodbye Nurse

- Questlines to deal with Nurse Sprocket once and for all.
- Over 18k words and 80 additional passages.

## Minor

- New fatigue-increasing drug.
- 25 Additional discussion topics on various topics.
- 3.5k words and 27 additional passages.

### Tester and Contributor Reward Scenes (designed as rewards for individual testers/backers, playable by all)

- VIP Scene: Club Egg.
- VIP Scene: Welcome To F.E.M.D.O.M.
- VIP Scene: Inviting A Wizard To F.E.M.D.O.M.
- VIP Scene: Going To Seed.
- VIP Scene: Dog On "Guard."
- VIP Scene: Software That Goes Hard.
- VIP Scene: Clean Feet, Dirty Minds.
- VIP Scene: Moo Moo More.
- VIP Scene: Secretary Inversion.
- Tester Scene: Slick, Shiny Slime.
- Tester Scene: The Most Arousing Game.
- Tester Scene: Orgasm Of The Species.
- Tester Scene: Galactic Convergence.
- Over 16.1k words and 101 new passages

Total Release Additions: 50.1k words/250 passages/86 achievements
Total Gamesize: 3.03 million words/11270 passages/5001 achievements
",
    "version": "1.0.3.2",
  },
  {
    "note": "
# 1.0.4.0 "Running Out of Witty Titles"

## 2 Fast 2 Forward

- Able to fast-forward from middle of day, rather than just from morning.

## The Locktober Revolution

- New late-game Locktober chastity event.
- Option to put Serena in chastity after she has been collared.
- Over 15k words and 81 new passages

## An Outfit On The Company

- New quest to diversify your office wardrobe thanks to a clothing voucher.
- Four new outfits as selected by backers.
- Over 4.9k words and 34 new passages

## Porn For the Porn Gods!

- Cum themed porn shoot (as chosen by backers!).
- Maid themed porn shoot (as chosen by backers!).
- Over 2.7k words and 16 new passages

## Minor

- HR Harassment Scene.
- Additional discussion themes on various topics.
- Option to ask Lisa Soto for some of her milk.

### Tester and Contributor Reward Scenes (designed as rewards for individual testers/backers, playable by all)

- Backer Scene: The Reward For Flexibility.
- Backer Scene: Giving 450%.
- VIP Scene: Only In Your Mind.
- VIP Scene: Wake Up And Apologise.
- VIP Scene: Just A Joke.
- VIP Scene: Tucked In... Personally.
- VIP Scene: Train Facts.
- VIP Scene: F.E.M.D.O.M.S.: Recruiting A Friend For Empress Strawberry.
- VIP Scene: Gothin' Maid: Meet Charly.
- VIP Scene: Milgram, I Hardly Knew Her!
- VIP Scene: Eating With An Egg.
- VIP Scene: Striped Socks On Camera.
- Tester Scene: Does She Have To Do Everything?
- Tester Scene: VERY strictly A-SLUT.
- Tester Scene: What's Better Than A Monorail?
- Tester Scene: A Roll Before A Roll In The Hay.
- Reward Scene: Aggressive Nursing.
- Over 31.5k words and 135 new passages

Total Release Additions: 64.2k words/331 passages/69 achievements
Total Gamesize: 3.1 million words/11611 passages/5073 achievements
",
    "version": "1.0.4.2",
  },
]
`;

exports[`DefaultGameInfoLoader > instance > should load artifact info from the meta folder 1`] = `
{
  "platform": "html",
  "version": "1.0.4.2",
}
`;

exports[`DefaultGameInfoLoader > instance > should load game info from the meta folder 1`] = `
{
  "authors": [
    {
      "metadataVersion": 0,
      "name": "deedee",
      "tfgamesSiteProfileId": 38045,
      "uuid": Uint8Array [
        243,
        179,
        179,
        180,
        27,
        27,
        75,
        59,
        139,
        59,
        59,
        27,
        27,
        59,
        27,
        27,
      ],
    },
  ],
  "description": "
![Cover](./startscreen.png)

\\[[Secretary Discord Server](https://discord.gg/XShMmqr)\\]

\\[[Play Game](http://www.secretarygame.com/game/index.html)\\]

\\[[Download the Android Version](https://mega.nz/folder/p6ImwZqZ#lm0fdhu_ndRPdkVMrf2haQ)\\]

\\[_Secretary_\\] is an erotic adventure/life simulation game, where the character starts as a normal guy in his late twenties who is about to transition from a dead end IT Job to a new position as the personal Secretary to a very important executive. Your choices choose how your development progresses as you interact with coworkers and friends and steer your way to learn that there's more going on at this company than might first appear...

Current fetishes include:

- (Forced) Feminization
- S&M
- Submission
- Corruption
- Chastity play
- Physical Transformation
- Humiliation
- Hand holding
- Bondage
- Bimbofication
- Validation
- Tentacles
- Latex

The game is in active development but is currently over a million words (longer than the collected words of shakespeare) with dozens of characters with unique art, a fully functioning inventory and nearly 600 pieces of equippable items.

Which content you experience is heavily controlled by player choice to range ranging from forced bimbofication to a (somewhat kinky) fully voluntary transition. In addition an expansive fetish flag system allows the player can tweak their experience in regards to what type of content they will encounter.

Gameplay information:

- The gameplay is primarily point and click with some inspiration taken from classic 90s adventure games, it also incorporates a fully functioning inventory, a variety of characters to get to know (and some to KNOW), as well as time advancement and random events.
- We release new content every other month (odd numbered months), each content release is at least 60k words, as well as new art and in-game items. Patrons/substars can gain early access to test builds, as well as help steer the topics of releases and some aspects of prioritizing development of the game.
- All saves should be forward compatible to new versions, if you want to play it safe when you change to a newer version save to disc, then load that save after updating to the new version.
- If you get stuck the game has a [wiki](https://sites.google.com/view/secretary-game-wiki/wiki/table-of-contents) (while it will contain full walkthroughs for each new version we are still in the process of backfilling some older content), you can also ask for help on forums or discord.

## IMPORTANT!

Please check #bugs on the Secretary Discord Server before posting any new bugs. We have regular updates there and maybe your problem has already been taken care of!

## 1.0.1.0 Obligatory DLC features:

### Dames: WIP

- Custom female art for all previous male-default characters.
- Kurt, Dr. Twig, Mr. Horn, Mr. Bergmeister, Mr. Bogart, Mr. Bogart, Dwayne Rockson, Stan, and Fabio each have new and unique feminine art.

### Lusts:

- Forced feminization porn shoot.
- Additional Delta Orgy event.
- Latex themed porn shoot (as chosen by backers!).
- Over 4.1k words and 24 new passages

### Collars:

- Option to multi-collar between Mrs. Antlers, Kurt, Bob, Alex, rather than only one collar per playthrough.
- Opportunity to upgrade from a company collar to Mrs. Antlers.
- Event where you can collar Serena.
- Option to show off your dominance skills to Mrs. Antlers with Serena or Wilco as your sub.
- Over 10k words and 58 new passages

### Minor Additions:

- NPC discussion themes refactored (backer voted QoL improvement).
- Additional ending handling for Serena.
- New "martial arts" drug that allows bypassing certain fighting skills scene requirements.

### Tester and Contributor Reward Scenes (designed as rewards for individual testers/backers, playable by all):

- Backer Scene: Mrs. Antlers Couple Therapy (normal and cruel variants).
- VIP Scene: Witnessing A Buy-Out.
- VIP Scene: Rooty For Your Booty.
- VIP Scene: Sissy On Board.
- VIP Scene: Managing Up.
- VIP Scene: Managing Down.
- VIP Scene: The Art Of Bondage.
- Tester Scene: Finally Out.
- Tester Scene: Ponies On Parade.
- Tester Scene: Shared Maintenance.
- Tester Scene: Not The Vend!
- Tester Scene: Worse Than Doxed?
- Tester Scene: Train On A Train.
- Over 25.2k words and 133 new passage

**Total Release Additions:** 50k words/248 passages/48 achievements
**Total Gamesize:** 2.93 million words/10757 passages/4828 achievements

New Inventory has been added but is still being updated as we speak.

The game is not yet finished. Please take a look at the current build and changelog to see whats new!
If the download needs a password it is always the gamename, written in lower case.
",
  "ids": {
    "tfgamesSiteGameId": 1445,
    "uuid": Uint8Array [
      158,
      9,
      117,
      21,
      239,
      129,
      70,
      17,
      165,
      48,
      236,
      143,
      71,
      96,
      51,
      169,
    ],
  },
  "metadataVersion": 0,
  "name": "Secretary",
  "synopsis": "An erotic adventure/life simulation game where the character starts as a normal guy in his late twenties who is about to transition from a dead end IT Job to a new position as the personal Secretary to a very important executive. Your choices choose how your development progresses as you interact with coworkers and friends and steer your way to learn that there's more going on at this company than might first appear...
",
  "tags": [
    "AT:MF",
    "AT:FF",
    "AT:MM",
    "AT:F-Self",
    "AT:M-Self",
    "AT:Sub",
    "AT:Dom",
    "TF:BE",
    "TF:Grow",
    "TF:Invol",
    "TF:M2F",
    "TF:Fem",
    "TF:Slow",
    "TF:MC",
    "TF:Bimbo",
    "TF:Herm",
    "TF:Corr",
    "TF:SheM",
    "TF:Vol",
    "TF:Sissy",
    "MM:2-D",
    "MM:Rend",
    "MM:Draw",
    "MM:SFX",
    "MM:Video",
  ],
}
`;
